File: test/python-tests/test_data_structures.py
# Test various data structures

# Lists
print("=== Lists ===")
fruits = ["apple", "banana", "cherry", "date"]
print(f"Original list: {fruits}")

# List operations
fruits.append("elderberry")
fruits.insert(1, "blueberry")
fruits.remove("cherry")
popped = fruits.pop()
print(f"After operations: {fruits}")
print(f"Popped item: {popped}")

# List slicing
print(f"First three: {fruits[:3]}")
print(f"Last two: {fruits[-2:]}")

# List methods
numbers = [3, 1, 4, 1, 5, 9, 2, 6]
numbers.sort()
print(f"Sorted numbers: {numbers}")
numbers.reverse()
print(f"Reversed numbers: {numbers}")
print(f"Count of 1: {numbers.count(1)}")
print(f"Index of 9: {numbers.index(9)}")

# Tuples
print("\n=== Tuples ===")
coordinates = (10, 20)
point = (3, 4, 5)
print(f"Coordinates: {coordinates}")
print(f"Point: {point}")

# Tuple unpacking
x, y = coordinates
print(f"Unpacked: x={x}, y={y}")

# Sets
print("\n=== Sets ===")
set1 = {1, 2, 3, 4, 5}
set2 = {4, 5, 6, 7, 8}
print(f"Set 1: {set1}")
print(f"Set 2: {set2}")

# Set operations
print(f"Union: {set1 | set2}")
print(f"Intersection: {set1 & set2}")
print(f"Difference: {set1 - set2}")
print(f"Symmetric difference: {set1 ^ set2}")

# Set methods
set1.add(6)
set1.remove(2)
print(f"Modified set 1: {set1}")
print(f"Is 3 in set1? {3 in set1}")
print(f"Set length: {len(set1)}")

# Dictionaries (extended)
print("\n=== Dictionaries Extended ===")
student = {
    "name": "Alice",
    "age": 20,
    "grades": [85, 90, 78, 92],
    "courses": {
        "math": "A",
        "science": "B+",
        "history": "A-"
    }
}

print(f"Student: {student}")
print(f"Keys: {list(student.keys())}")
print(f"Values: {list(student.values())}")

# Nested dictionary access
print(f"Math grade: {student['courses']['math']}")

# Dictionary methods
student.update({"major": "Computer Science"})
age = student.pop("age")
print(f"After update and pop: {student}")
print(f"Popped age: {age}")

# Default dictionary
grades = student.get("grades", [])
print(f"Grades: {grades}")
missing = student.get("missing_key", "default_value")
print(f"Missing key with default: {missing}")

# Complex nested structure
complex_data = {
    "users": [
        {"id": 1, "name": "Alice", "active": True},
        {"id": 2, "name": "Bob", "active": False},
        {"id": 3, "name": "Charlie", "active": True}
    ],
    "metadata": {
        "total_users": 3,
        "active_users": 2,
        "created_date": "2024-01-01"
    }
}

print(f"\nComplex data: {complex_data}")
print(f"Active users: {[u['name'] for u in complex_data['users'] if u['active']]}")

File: test/python-tests/test_data_structures.cpp
#include <iostream>
#include <string>
#include <vector>
#include <optional>
#include <unordered_map>
#include <tuple>
#include <utility>
#include <type_traits>
#include <functional>
#include <sstream>
#include <iomanip>
#include <cmath>
#include <algorithm>
#include <typeinfo>
#include <cstring>

using namespace std;

// helpers
template<typename T> static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v);
template<typename T> static inline std::string to_str(const T& v){ std::ostringstream os; os<<v; return os.str(); }
static inline std::string to_str(const std::string& s){ return s; }
static inline std::string to_str(const char* s){ return std::string(s); }
static inline std::string to_str(double v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
static inline std::string to_str(float v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
template<typename T> static inline int to_int(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stoi(v); else return static_cast<int>(v); }
template<typename T> static inline float to_float(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stof(v); else return static_cast<float>(v); }
template<typename C> static inline int sum(const C& c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline int sum(std::initializer_list<T> c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline size_t len(const T& c){ return c.size(); }
static inline size_t len(const char* s){ return std::char_traits<char>::length(s); }
template<typename K, typename V> static inline std::string to_str(const std::unordered_map<K,V>&){ return std::string("{...}"); }
// minimal ABC base to support 'class X : public ABC'
struct ABC {};
// minimal math namespace replacement for 'math.pi'
struct __fluxus_math { static constexpr double pi = 3.141592653589793; static inline double sqrt(double x){ return std::sqrt(x); } } math;
// minimal sys module stub
namespace sys { static std::vector<std::string> argv; }
namespace asyncio { template<typename F> static inline void run(F f){ f(); } }

// vector to string conversion for printing
template<typename T>
static inline std::string vec_to_str(const std::vector<T>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        if constexpr (std::is_same_v<T, std::string>) {
            os << "'" << v[i] << "'";
        } else {
            os << v[i];
        }
    }
    os << "]";
    return os.str();
}

// nested vector to string
template<typename T>
static inline std::string vec_to_str(const std::vector<std::vector<T>>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        os << vec_to_str(v[i]);
    }
    os << "]";
    return os.str();
}

// stream operator for vectors
template<typename T>
static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v) {
    os << vec_to_str(v);
    return os;
}
template<typename K, typename V> static inline std::ostream& operator<<(std::ostream& os, const std::unordered_map<K,V>&){ os << "{...}"; return os; }
template<typename K, typename V> static inline auto py_items(const std::unordered_map<K,V>& m){ std::vector<std::pair<K,V>> v; v.reserve(m.size()); for(const auto& kv: m){ v.emplace_back(kv.first, kv.second);} return v; }


int main() {
    std::cout << "=== Lists ===" << std::endl;
    auto fruits = std::vector<std::string>{"apple", "banana", "cherry", "date"};
    std::cout << to_str("Original list: ") << to_str(fruits) << std::endl;
    fruits.push_back("elderberry");
    fruits.insert(1, "blueberry");
    fruits.remove("cherry");
    auto popped = fruits.pop();
    std::cout << to_str("After operations: ") << to_str(fruits) << std::endl;
    std::cout << to_str("Popped item: ") << to_str(popped) << std::endl;
    std::cout << to_str("First three: ") << to_str(fruits[) << std::endl;
    std::cout << to_str("Last two: ") << to_str(fruits[-2) << std::endl;
    auto numbers = std::vector<decltype(3)>{3, 1, 4, 1, 5, 9, 2, 6};
    numbers.sort();
    std::cout << to_str("Sorted numbers: ") << to_str(numbers) << std::endl;
    numbers.reverse();
    std::cout << to_str("Reversed numbers: ") << to_str(numbers) << std::endl;
    std::cout << to_str("Count of 1: ") << to_str(numbers.count(1)) << std::endl;
    std::cout << to_str("Index of 9: ") << to_str(numbers.index(9)) << std::endl;
    std::cout << "\n=== Tuples ===" << std::endl;
    auto coordinates = (10, 20);
    auto point = (3, 4, 5);
    std::cout << to_str("Coordinates: ") << to_str(coordinates) << std::endl;
    std::cout << to_str("Point: ") << to_str(point) << std::endl;
    auto x, y = coordinates;
    std::cout << to_str("Unpacked: x=") << to_str(x) << to_str(", y=") << to_str(y) << std::endl;
    std::cout << "\n=== Sets ===" << std::endl;
    auto set1 = {1, 2, 3, 4, 5};
    auto set2 = {4, 5, 6, 7, 8};
    std::cout << to_str("Set 1: ") << to_str(set1) << std::endl;
    std::cout << to_str("Set 2: ") << to_str(set2) << std::endl;
    std::cout << to_str("Union: ") << to_str(set1 | set2) << std::endl;
    std::cout << to_str("Intersection: ") << to_str(set1 & set2) << std::endl;
    std::cout << to_str("Difference: ") << to_str(set1 - set2) << std::endl;
    std::cout << to_str("Symmetric difference: ") << to_str(set1 ^ set2) << std::endl;
    set1.add(6);
    set1.remove(2);
    std::cout << to_str("Modified set 1: ") << to_str(set1) << std::endl;
    std::cout << to_str("Is 3 in set1? ") << to_str(3 in set1) << std::endl;
    std::cout << to_str("Set length: ") << to_str(len(set1)) << std::endl;
    std::cout << "\n=== Dictionaries Extended ===" << std::endl;
    auto student = {;
    };
    std::cout << to_str("Student: ") << to_str(student) << std::endl;
    std::cout << to_str("Keys: ") << to_str(list(student.keys())) << std::endl;
    std::cout << to_str("Values: ") << to_str(list(student.values())) << std::endl;
    std::cout << to_str("Math grade: ") << to_str(student['courses']['math']) << std::endl;
    student.update({"major": "Computer Science"});
    auto age = student.pop("age");
    std::cout << to_str("After update and pop: ") << to_str(student) << std::endl;
    std::cout << to_str("Popped age: ") << to_str(age) << std::endl;
    auto grades = student.get("grades", []);
    std::cout << to_str("Grades: ") << to_str(grades) << std::endl;
    auto missing = student.get("missing_key", "default_value");
    std::cout << to_str("Missing key with default: ") << to_str(missing) << std::endl;
    auto complex_data = {;
    };
    std::cout << to_str("\\nComplex data: ") << to_str(complex_data) << std::endl;
    std::cout << to_str("Active users: ") << to_str(0) << std::endl;
    return 0;
}

